rfcs/0011-separate-flyway-from-pipeline.md

```diff
@@ -0,0 +1,389 @@
+* Typical deployment option for almost all changes (migration or otherwise)
+* New applications are deployed with a period of overlap with the old applications
+* Requires no downtime
+* In-place deployment that requires no additional hardware
+* Has constraints on the type of migrations that can be deployed (discussed in detail below)
+
+##### Abbreviated Deployment Diagram
+
+![2022-02 Deployment Sequence](./resources/0011-2022-02-deployment-sequence.svg)
+
+<details><summary>Diagram Notes</summary>
+
+As with most graphical representations, it can be difficult to strike the appropriate balance between information density and accuracy. The `Jenkins` participant is shorthand for our existing "BFD - Multibranch and Multistage Pipeline." Participants like `bfd-pipeline` and `bfd-server` are composites of the Jenkins deployment stage (generalized for the three existing `test`, `prod-sbx`, and `prod` environments), necessary AWS API endpoints that accommodate the deployment of these resources via terraform, **and** the resultant BFD resources running in each environment.
+
+</details>
+
+
+#### Cloned deployment
+* A cloned environment handles traffic while the primary instance undergoes the deployment and then traffic is
+redirected back to the primary
+* Manual deployment
+* Requires no downtime
+* Has fewer constraints on the types of migrations that can be deployed than the standard Jenkins deployment
+
+#### Downtime deployment
+* Service is interrupted for some period of time
+* Has additional coordination and communication requirements and acceptance of downtime window and risk mitigation plan
+* Manual or Automated deploy are possible during the downtime window
+* Can be used when downtime is required or desired due to other factors
+
+In general, a Jenkins deployment is preferred because it is a fully automated process that results in no
+downtime, has low risk of human error, no additional hardware costs, and requires no coordination or communication
+apart from what is otherwise required for any particular change. The other types of deployments can be used for
+special situations where it is determined to be preferable for reasons of cost, risk mitigation, or otherwise. This RFC
+will focus on optimizing the Standard Jenkins deployment since it is the preferred deployment and is most commonly used.
+Optimizing the Jenkins deployment can also make it a more viable option for certain complex migrations that otherwise
+would be candidates for a cloned deployment or a downtime deployment.
+
+### Constraints on migrations that are deployed via Jenkins deployment
+
+During a Jenkins deployment the old API server continues to run and serve traffic during the deployment of the new
+applications until the new API server is fully deployed and available. This means that it is possible that the old API
+server will attempt to read from the database after the new schema is in place. This leads to a constraint on BFD
+migrations:
```

Comment:
Perhaps a strange thought, but this is creating a constraint because we can't just clone the DB and bring up a new DB instance with all the data in the new schema configuration, right? Upon reading this, my initial naive thought was "we clone the BFD servers to maintain uptime, why not clone the new DB too and delete the old one once its BFD servers are replaced fully?" I assume it's because the data contained is too massive to clone, and would be prohibitive on a time/money scale, so its not a consideration here.

Not sure if that point is worth mentioning for completeness; but the rest of the RFC seems to explicitly spell out many parts of the migration strategy, so maybe?

Reply:
We can do that but it is a different type of deployment. I've referred to it above as a cloned deployment and it does have fewer constraints but has additional costs -- additional hardware but also a window where writes would have to be made to both clusters (which becomes more relevant when RDA goes live), and is currently not an automated solution (although we would like to get there eventually which would make that concern less important).